include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,211 @@
+        return reinterpret_cast<char*>(aligned_allocator_t(a).allocate(n * ceil_div(size, align)));
+    }
+
+    template <std::size_t N, class Allocator>
+    typename std::enable_if<N < alignof(max_align_t), char*>::type
+    try_allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align) {
+        if (align <= N) {
+            using aligned_t = typename boost::type_with_alignment<N>::type;
+            using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+            return reinterpret_cast<char*>(aligned_allocator_t(a).allocate(n * ceil_div(size, align)));
+        } else {
+            return try_allocate_aligned<2*N>(a, n, size, align);
+        }
+    }
+
+    template <class Allocator>
+    char* allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align) {
+        return try_allocate_aligned<1>(a, n, size, align);
+    }
+
+    template <std::size_t N, class Allocator>
+    typename std::enable_if<alignof(max_align_t) <= N, void>::type
+    try_deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+        using aligned_t = max_align_t;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+    }
+
+    template <std::size_t N, class Allocator>
+    typename std::enable_if<N < alignof(max_align_t), void>::type
+    try_deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+        if (align <= N) {
+            using aligned_t = typename boost::type_with_alignment<N>::type;
+            using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+            aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+        } else {
+            try_deallocate_aligned<2*N>(a, p, n, size, align);
+        }
+    }
+
+    template <class Allocator>
+    void deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+        try_deallocate_aligned<1>(a, p, n, size, align);
+    }
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+    using impl_allocator_type = boost::allocator_rebind_t<Allocator, impl>;
+
+    class impl : public detail::any_allocator_base
+    {
+         char_allocator_type a_;
+
+    public:
+        explicit
+        impl(char_allocator_type const& a)
+        : any_allocator_base(typeid(char_allocator_type)), a_(a)
+        {
+        }
+
+        char*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            if (size == 1) {
```

Comment:
why the special case?

Reply:
Because we don't need to rebind.

Reply:
is this an optimization?

Reply:
Yes. Peter suggested it. We store an allocator to char since we have to store something and try to use it without rebinding by default. Rebinding might have a small cost, especially if the allocator is stateful. Even for simple allocators like pmr.

Reply:
That's not what Peter said, you need to take this condition out